test/jdk/javax/swing/JButton/bug4323121.java

```diff
@@ -30,50 +30,80 @@
  * @run main bug4323121
  */
 
-import java.awt.Graphics;
 import java.awt.Point;
 import java.awt.Robot;
+import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
-import java.awt.event.MouseListener;
-import java.awt.event.MouseMotionListener;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.util.concurrent.CountDownLatch;
+
 import javax.swing.JButton;
 import javax.swing.JFrame;
 import javax.swing.SwingUtilities;
 
-public class bug4323121 {
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class bug4323121 extends MouseAdapter {
 
     static JFrame frame;
-    static testButton button;
-    static volatile Point pt;
-    static volatile int buttonW;
-    static volatile int buttonH;
-    static volatile boolean failed = false;
+    static JButton button;
+
+    static volatile Point buttonCenter;
+
+    private static final CountDownLatch windowGainedFocus = new CountDownLatch(1);
+
+    private static final CountDownLatch mouseEntered = new CountDownLatch(1);
+
+    // Thread-safe by using the mouseEntered latch
+    private static boolean modelArmed;
 
     public static void main(String[] args) throws Exception {
         Robot robot = new Robot();
         robot.setAutoDelay(100);
+
+        final bug4323121 eventHandler = new bug4323121();
         try {
             SwingUtilities.invokeAndWait(() -> {
+                button = new TestButton("gotcha");
+                button.addMouseMotionListener(eventHandler);
```

Comment:
I just kept it because it was there. When the event handler extends `MouseAdapter`, there are no additional methods to override.

However, you're right, `MouseMotionListener` is not needed.

Reply:
Since you are simplifying the test, I guess this can be considered.